### ISSUE TITLE
RBAC is allocating a lot of memory

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#40015]
I was looking into why kubemark-scale is failing and looked into memory allocations. It seems that RBAC is responsible for significant portion of it:
```
(pprof) wojtekt@wojtekt-work:~/gke-scripts/google3/cloud/kubernetes$ go tool pprof --alloc_space http://localhost:12345/debug/pprof/heap
Fetching profile from http://localhost:12345/debug/pprof/heap
Saved profile in /usr/local/google/home/wojtekt/pprof/pprof.localhost:12345.alloc_objects.alloc_space.010.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) top
367.43GB of 739.62GB total (49.68%)
Dropped 1252 nodes (cum <= 3.70GB)
Showing top 10 nodes out of 272 (cum >= 11.85GB)
      flat  flat%   sum%        cum   cum%
  116.08GB 15.69% 15.69%   129.35GB 17.49%  net/http.(*http2serverConn).newWriterAndRequest
   54.22GB  7.33% 23.02%    92.37GB 12.49%  k8s.io/kubernetes/pkg/registry/rbac/validation.(*DefaultRuleResolver).RulesFor
   43.72GB  5.91% 28.94%    43.72GB  5.91%  math/big.nat.mul
   37.60GB  5.08% 34.02%    37.60GB  5.08%  reflect.unsafe_New
   24.75GB  3.35% 37.37%    35.29GB  4.77%  math/big.nat.divLarge
   21.83GB  2.95% 40.32%    21.83GB  2.95%  k8s.io/kubernetes/pkg/registry/rbac/validation.appliesToUser
   20.44GB  2.76% 43.08%    20.44GB  2.76%  k8s.io/kubernetes/pkg/api.DeepCopy_api_NodeStatus
   18.80GB  2.54% 45.62%    27.47GB  3.71%  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/protobuf.(*Serializer).Encode
   18.14GB  2.45% 48.08%    18.14GB  2.45%  bytes.makeSlice
   11.85GB  1.60% 49.68%    11.85GB  1.60%  k8s.io/kubernetes/pkg/client/cache.(*threadSafeMap).List
```

I think that some of it might be related to the fact the we still have incorrect role for kube-proxy in kubemark that is not allowed to list endpoints, but that should hopefully be fixed soon - @shyamjvs 

Once this is fixed, we should verify where we are, and potentially optimize RBAC-related code.

@kubernetes/sig-scalability-misc @shyamjvs @deads2k @liggitt

**Comment by shyamjvs** [comment]
@wojtek-t Thanks a lot for finding this out.
@deads2k One question I have in this regard is that, do we send the certificate in the header for every API request when using RBAC? I hope that we have something better than that as it could be a real network and computation overhead. On the other hand, not sending certificate on each request would mean we need to maintain the connection state with each of the client (which I guess is not a possible solution). 
- On the other hand, this might not be related to the authentication step at all and the issue lies in the authz (as kubemark-scale was working back in the ABAC times). IIUC the issue in authz is because of writing snd reading these roles and bindings through API requests (while previously it was just a file on the master). How about we have a separate storage cache for authz related API requests (as we know that all API requests have to definitely go through authz and this cache would hardly have to refresh (as most of the roles and bindings are written mostly during the cluster creation))?

**Comment by deads2k** [comment]
> How about we have a separate storage cache for authz related API requests (as we know that all API requests have to definitely go through authz and this cache would hardly have to refresh (as most of the roles and bindings are written mostly during the cluster creation))?

The RBAC authorizer operates out of a cache, so during the authorization step it does not go to and from the API server.

`k8s.io/kubernetes/pkg/registry/rbac/validation.(*DefaultRuleResolver).RulesFor` seems to suggest authz over authn.

**Comment by shyamjvs** [comment]
> The RBAC authorizer operates out of a cache, so during the authorization step it does not go to and from the API server.

Yes. But my question is if there is a separate dedicated cache for just the RBAC related requests or we just use the regular apiserver cache?

**Comment by deads2k** [comment]
> Yes. But my question is if there is a separate dedicated cache for just the RBAC related requests or we just use the regular apiserver cache?

Dedicated.  It only holds the latest state (unlike the watch cache) and it never discards an element (unlike the deserialization cache).  Is there another cache I'm unaware of?

**Comment by shyamjvs** [comment]
No that's fine. I was indeed talking about the watch cache.

**Comment by liggitt** [comment]
> do we send the certificate in the header for every API request when using RBAC? I hope that we have something better than that as it could be a real network and computation overhead. On the other hand, not sending certificate on each request would mean we need to maintain the connection state with each of the client (which I guess is not a possible solution).

certificates are not sent in headers, they are sent during TLS connection establishment. that means that keep-alive connections only authenticate certificates when the connection is first set up.

bearer tokens are sent in headers, which are sent with every request, and verified with every request.

**Comment by shyamjvs** [comment]
@liggitt This might be a basic question, but does it mean that we should prefer using bearer tokens (or similar mechanisms) more than certificates, as using more certificates would lead to the apiserver maintaining more TLS connections? Or on the contrary, using bearer tokens is more expensive computationally since we need to verify each API request? If we know that one of them is better for sure, we should prefer using that mode of authentication wherever possible.

**Comment by liggitt** [comment]
> using more certificates would lead to the apiserver maintaining more TLS connections

keep-alive is not determined by cert usage. the restclient does keep-alive even without cert usage, since network and TLS-establishment is expensive.

> Or on the contrary, using bearer tokens is more expensive computationally since we need to verify each API request?

Both have benefits and costs. Certs establish once, but are more difficult to manage (requires PKI to create, not possible to revoke). Tokens are verified per request, but are easier to create, can be revoked, and are more flexible in the number of systems they can integrate with.

> If we know that one of them is better for sure, we should prefer using that mode of authentication wherever possible.

Node bootstrapping sets up client certs for nodes, so in most deployments, nodes use client certs.

**Comment by shyamjvs** [comment]
That clarifies a lot of questions. Thank you @liggitt :)

**Comment by wojtek-t** [comment]
After fixing the roles for hollow kube-proxy, it seems much better:
```
  246.62GB 19.53% 19.53%   252.69GB 20.01%  net/http.(*http2serverConn).newWriterAndRequest
  136.76GB 10.83% 30.36%   136.76GB 10.83%  reflect.unsafe_New
   54.30GB  4.30% 34.66%    73.52GB  5.82%  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/protobuf.(*Serializer).Encode
   45.54GB  3.61% 38.27%    45.54GB  3.61%  k8s.io/kubernetes/pkg/api.DeepCopy_api_NodeStatus
   39.48GB  3.13% 41.39%    39.48GB  3.13%  bytes.makeSlice
   37.26GB  2.95% 44.34%    37.76GB  2.99%  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/protobuf.(*RawSerializer).Encode
   31.61GB  2.50% 46.85%    31.61GB  2.50%  k8s.io/kubernetes/pkg/storage.newCacheWatcher
   25.92GB  2.05% 48.90%    25.92GB  2.05%  k8s.io/kubernetes/pkg/client/cache.(*threadSafeMap).List
   23.43GB  1.86% 50.75%    40.89GB  3.24%  k8s.io/kubernetes/pkg/registry/rbac/validation.(*DefaultRuleResolver).RulesFor
```

Though 23GB used by ``` RulesFor ``` still seems quite a lot...

**Comment by wojtek-t** [comment]
@shyamjvs - if you have free cycles, maybe you can look into this (i.e. why rbac validation is allocating tens of GB of memory).

**Comment by wojtek-t** [comment]
My feeling is that we may be doing some unnecessary copies somewhere, and we may save a lot just from some simple change.

**Comment by shyamjvs** [comment]
@wojtek-t Sure, I can look into this.

**Comment by shyamjvs** [comment]
@deads2k It turns out that the following step in the [RulesFor function](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/rbac/validation/rule.go#L96) is quite wasteful:
`if clusterRoleBindings, err := r.clusterRoleBindingLister.ListClusterRoleBindings(); err != nil {`
As you anyway filter for only those rules that are applicable to a particular user (which hopefully would be just a few), there is no need for fetching the entire array of `clusterRoleBindings`.
I suggest that we (add a new one or) modify function `ListClusterRoleBindings()` to `ListClusterRoleBindings(user user.Info)` to return only the bindings relevant to `user`.
This should greatly reduce the memory consumption by rbac. If you agree with this idea, I can create a PR which brings this change.

**Comment by deads2k** [comment]
You think the slice of pointers (and small ones at that) is the memory allocation problem?  I'm ok with using a filtering call (not actually taking the user info, but its constituent subjects), but it would be good to know just how much we gain.

**Comment by wojtek-t** [comment]
Is this ListClusterRoleBinding() reaching registry are those these "StaticRoles" ? If the latter, I would be really surprised if it is this, as we are not even copying anything there.
Also - this method is returning pointers, so we potentially shouldn't even copy anything.

**Comment by wojtek-t** [comment]
Are you sure this in fact is not a problem of making copies when passing args to ```GetRoleReferenceRules ``` and/or ```appliesTo ``` functions? And/or appending things to ```policyRules ```?

Do you have any profile to confirm that?

**Comment by shyamjvs** [comment]
> Are you sure this in fact is not a problem of making copies when passing args to GetRoleReferenceRules and/or appliesTo functions? And/or appending things to policyRules?

Both of the functions `appliesTo` and `GetRoleReferenceRules` are just called multiple times in a loop with small structs as arguments, so it would be surprising if they really consume so much memory.
And neither should `policyRules` be big for one user and that's the minimal memory we would anyway need to evaluate a given request (unless you pass some extra context to the RulesFor function at all places where it is called so that it just returns only those rules which are relevant at the place where they are used. This IIUC would make the code more complex.)

**Comment by shyamjvs** [comment]
Sorry if I make some mistakes in drawing these inferences as I'm still new to golang :)

**Comment by wojtek-t** [comment]
>Both of the functions appliesTo and GetRoleReferenceRules are just called multiple times in a loop with small structs as arguments, so it would be surprising if they really consume so much memory.
And neither should policyRules be big for one user and that's the minimal memory we would anyway need to evaluate a given request (unless you pass some extra context to the RulesFor function at all places where it is called so that it just returns only those rules which are relevant at the place where they are used. This IIUC would make the code more complex.)

Even if they are small, we are having thousands of QPSes in these tests. And we are calling this function for every request. So I wouldn't be surprised if these copies are actually suming up to these tens of GBs.

**Comment by shyamjvs** [comment]
> Even if they are small, we are having thousands of QPSes in these tests. And we are calling this function for every request. So I wouldn't be surprised if these copies are actually suming up to these tens of GBs.

Yes that makes sense. Would it be good to save those extra bytes in the function calls at the cost of increased code complexity (though it's possible that I'm missing a better way around here and more profiling might be needed).

**Comment by wojtek-t** [comment]
What do you mean by "increase complexity"? If I am indeed right, we should just pass pointers instead of structs and that would do the job. But we need to verify if I am in fact right.

**Comment by shyamjvs** [comment]
Ah! I get it now. Sorry, I was referring to this in my last comment:
> unless you pass some extra context to the RulesFor function at all places where it is called so that it just returns only those rules which are relevant at the place where they are used. This IIUC would make the code more complex.

So IIUC then changing from `[]rbac.PolicyRule` -> `[]*rbac.PolicyRule` wherever we can (and at most places we can) should reduce array copying across function calls.

**Comment by wojtek-t** [comment]
It should - though we need to ensure that this is safe (i.e. that noone will modify the pointers in the background).

**Comment by ericchiang** [comment]
Did anyone take this? I've got a couple cycles to work on RBAC related things for this release.

**Comment by wojtek-t** [comment]
@ericchiang - from what I know noone didn't take closer look. If you have some cycle, it would be great if you could take a look. Thanks!

**Comment by timothysc** [comment]
per: @kubernetes/sig-scalability-bugs review if this is still an issue.

**Comment by smarterclayton** [comment]
Fairly sure we should just implement the same abstractions we added in openshift to avoid this excessive copying.

**Comment by liggitt** [comment]
switched `RulesFor` access pattern to a visitor that takes a `*rbac.PolicyRule` in https://github.com/kubernetes/kubernetes/pull/44449 (and added short-circuiting on success)

**Comment by liggitt** [comment]
closed by #44449
